### DISCUSSION TITLE
IKE SA states on receiving IKE_DOWN event

### DISCUSSION THREAD

**Discussion by hbhokare** [strongswan/strongswan#525] [Q&A]
Hi,
We are trying to develop the IPSec based application and using different strongswan/davici API from our application. We have following questions related to event handling.
1. What are different ike sa states as provided by strongswan on IKE_DOWN event. We wanted to check transition of ike sa states when we recevied IKE_DOWN event.
2. Currently we considered DELETING and DESTROYING  states. Can you please tell us all other states and what action strongswan is doing when ike sa is traversing through those states. 
Thanks in Advance !

**Comment by tobiasbrunner** [comment]
It won't be `IKE_DESTROYING` as that is only set shortly before an IKE_SA is actually destroyed (i.e. `ike_updown()` is always called before that if it's called at all). It also won't be `IKE_CREATED` and `IKE_CONNECTING` as there will only be a "down" event after there was an "up" event, which only happens once an SA reaches `IKE_ESTABLISHED` state. It also won't be  `IKE_REKEYED` as no "down" events are triggered for SAs in that state.

Other than that it could be any state if errors cause an `IKE_SA` to get locally destroyed without notifying the peer or if the deletion of an IKE_SA is forced and no DELETE can be sent because some other exchange is in progress. There won't be a state change before the call to `ike_updown()` in those cases. So while it usually will be `IKE_DELETING` (that's what happens during a proper delete), an IKE_SA could be in either `IKE_ESTABLISHED` or `IKE_REKEYING` state too when a "down" event is triggered (`IKE_PASSIVE` might also be a possibility if the _ha_ plugin is used).

Why is the state of the IKE_SA important to you during that event?

**Reply by hbhokare** [reply]
Thanks, Actually I am interested in IKE_SA state after I received IKE_DOWN based on which I can take a decision of deallocating memory reserved in application to store ike sa.

**Reply by tobiasbrunner** [reply]
How does that depend on the state? An IKE_SA for which a "down" event was triggered will never go "up" again, no matter what state it was in, so you can always deallocate any state you keep yourself when you receive such an event. Unless you already did so, you will also have to implement a listener for the _ike-rekey_ event as there won't be any _ike-updown_ events for SAs involved in rekeyings.

**Reply by hbhokare** [reply]
okay thanks !

**Comment by amvarunn** [comment]
Hi, @tobiasbrunner 
In continuation with the above question, please clarify some things related to CHILD_UPDOWN.
At what state do we get a CHILD_UP, and at what state do we get a CHILD_DOWN?
When we get an IKE_UP/DOWN, will it be in a deterministic state, or, is there a set of states where we may see them?

Listing the states which we are concerned about. 
   "CREATED", "ROUTED", "INSTALLING", "INSTALLED", "UPDATING", "REKEYING", "REKEYED", "RETRYING", "DELETING",  "DELETED", "DESTROYING"
   
   Thank you

**Reply by Pankaj-Deshpande-Com** [reply]
Hello @tobiasbrunner ,   

This is an urgent query. First, I am clarifying above query further so as to expedite your response and give you more background as to what we are trying to achieve and associated challenges which you can help us surmount.

- We found possible CHILD SA specific states getting reported to VICI application are : "CREATED", "ROUTED", "INSTALLING", "INSTALLED", "UPDATING", "REKEYING", "REKEYED", "RETRYING", "DELETING", "DELETED", "DESTROYING"

 - In our VICI API based application, we are doing only reference counting for child SAs rather than remembering the full identifiers etc - this is to save memory and make our design scalable.
 
 - We are doing ChildSARefCnt++ whenever we receive CHILD SA UP event  and We are doing ChildSARefCnt--  whenever we receive CHILD SA DOWN event.  But this is not correct as Strongswan might sent two UP events in creation path e.g. UP(CREATED)-->UP(INSTALLED) but in destroy path, it may give us single  DOWN (DELETED) event. And there may be many more such combinations in which ref counting can go wrong. 
 
 - We are keen to know what states we should use for reference counting in UP and DOWN path 
 e.g.Take reference when we receive CHILD SA with state=CREATED and Down the reference when we receive CHILD SA DOWN with state=DELETED ?

 - So What states reported in UP event and in what all states reported with DOWN event should we hook our reference count logic should we attach to ?
 and exactly how ? 
 
 You had helped us solve the issue related to IKE SA UP/DOWN events and now we need your urgent help on CHILD SA events... 
 
 Regards
Pankaj Deshpande
